### ISSUE TITLE
Use gccemacs to compile packages in parallel

### ISSUE THREAD

**Issue by jjzmajic** [raxod502/straight.el#492]
Something like `(native-compile-async "~/.emacs.d/straight/build" 4 t)` would be great. Plus it solves the problem of async compilation for people building from source. Once it lands in master (and it should around Emacs 28, the feature branch seems very mature) it will be the default anyway.

PS: [This](https://akrl.sdf.org/gccemacs.html) is the feature branch I am talking about.

**Comment by raxod502** [comment]
Just to make sure I am understanding:

* this feature provides a way to compile Elisp files to a form that can be executed more efficiently, so it's desirable to make it available as an option for `straight.el`
* although we could already implement asynchronous byte-compilation, using this feature would make that implementation simpler in `straight.el`

Is that right?

If so, I'd point out that there are two levels:

* byte-compile the files of a single package asynchronously, which is not a problem
* byte-compile the files of multiple packages in parallel, which requires the implementation of a new interface by which multiple packages can have their recipes specified at the same time; otherwise `straight.el` doesn't know ahead of time what packages will need to be built or what their recipes are

**Comment by alphapapa** [comment]
Basically, something like this:
```el
(use-package comp
  :config
  (defcustom native-compile-async-jobs
    (or (ignore-errors
	  (string-to-number (shell-command-to-string "nproc")))
	1)
    "How many jobs to use."
    :type 'integer)

  (setf native-compile-async-jobs 2)

  (defun ap/package-native-compile-async (package &optional all)
    "Compile PACKAGE natively, or with prefix ALL, all packages."
    (interactive (list (unless current-prefix-arg
			 (completing-read "Package: " (mapcar #'car package-alist)))))
    (let* ((directory (if package
			  (file-name-directory (locate-library package))
			package-user-dir)))
      (native-compile-async directory native-compile-async-jobs t)
      (load-library (symbol-name package)))))
```
However, adding such a feature to Straight is probably a little premature.  I hope the branch is merged for Emacs 28, but let's wait and see.  :)

**Comment by raxod502** [comment]
Is this solved by #507?

**Comment by jjzmajic** [comment]
Yes, sir!
